### PR TITLE
Adds StorageConsumer, supporting all storage except Elasticsearch AWS

### LINKED ISSUE #8: Add StorageConsumer
While we can temporarily copy/paste special-cased code for elasticsearch, I think we should be able to make an automatic StorageConsumer which works when any existing zipkin storage auto-configuration is present. This could go in the core sparkstreaming library and autoconfig since it would have no new deps. It might look like this..

```java
package zipkin.sparkstreaming;

import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import zipkin.Span;
import zipkin.storage.Callback;
import zipkin.storage.StorageComponent;

public abstract class StorageConsumer implements Consumer, Closeable {
  transient volatile StorageComponent instance;
  protected abstract StorageComponent tryCompute();

  @Override public final void accept(Iterable<Span> trace) {
    get().asyncSpanConsumer().accept(asList(trace), Callback.NOOP);
  }

  final StorageComponent get() {
    StorageComponent result = instance;
    if (result == null) {
      synchronized (this) {
        result = instance;
        if (result == null) {
          instance = result = tryCompute();
        }
      }
    }
    return result;
  }

  // TODO: not sure how spark will ever call this..
  @Override public final void close() throws IOException {
    synchronized (this) {
      if (instance != null) {
        instance.close();
      }
    }
  }

  static <E> List<E> asList(Iterable<E> iter) {
    List<E> list = new ArrayList<E>();
    for (E item : iter) list.add(item);
    return list;
  }
}
```

Then in autoconfiguration we'd enable when zipkin.storage.type is set. Then, somehow we collect the properties we later apply on-demand to recreate the storage component..

```java
class AutoConfiguredStorageConsumer extends StorageConsumer {
  final Properties zipkinStorageProperties;
  @Override protected abstract StorageComponent tryCompute() {
    AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
    addEnvironment(context, zipkinStorageProperties);
    context.register(PropertyPlaceholderAutoConfiguration.class,
        // wonder if we can figure out which classes to put here dynamically
    );
    context.refresh();
    try {
      return context.getBean(StorageComponent.class);
    } finally {
      context.close();
    }
  }
}
```

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This consumer writes spans to a Zipkin `StorageComponent`, such as
Elasticsearch.

In order to connect, you minimally need to set `zipkin.storage.type` to
a configured backend. You also need to set relevant properties for that.

Ex.
```bash
java -jar zipkin-sparkstreaming-job.jar \
  --zipkin.storage.type=elasticsearch \
  --zipkin.storage.elasticsearch.hosts=http://127.0.0.1:9200 \
  ...
```

Configuration properties can be set via commandline parameters, system
properties or any other alternative [supported by Spring Boot](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html).

Besides setting storage properties, you can also override the following.
All of the below have a prefix of "zipkin.sparkstreaming.consumer.storage"

Property | Default |Description
--- | --- | ---
fail-fast | true | check storage before submitting the job.

Fixes #8

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -47,7 +47,7 @@
     <scala.binary.version>2.10</scala.binary.version>
     <spark.version>1.6.3</spark.version>
 
-    <zipkin.version>1.20.0</zipkin.version>
+    <zipkin.version>1.20.1-SNAPSHOT</zipkin.version>
```
Comment (adriancole) [inline pom.xml]: pending release of code that makes storage property types serializable

File: sparkstreaming/src/test/java/zipkin/sparkstreaming/StorageConsumerTest.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+package zipkin.sparkstreaming;
+
+import org.junit.Test;
+
+public class StorageConsumerTest {
+
+  @Test
+  public void test() {
+    // TODO
```
Comment (adriancole) [inline sparkstreaming/src/test/java/zipkin/sparkstreaming/StorageConsumerTest.java]: I will write this.. currently tested manually both in dev mode and also to a spark cluster

**Comment by adriancole** [comment]
This now supports all storage except Elasticsearch AWS

**Comment by adriancole** [comment]
will rebase this on zipkin 1.20.1 once syncs to maven central

**Review by naoman — APPROVED** [review]
Looks good to me!

**Comment by adriancole** [comment]
ok backfilled tests as promised.

**Pull request merged**